src/OrleansAzureUtils/Providers/Storage/AzureTableStorage.cs

```diff
@@ -246,49 +250,112 @@ internal void ConvertToStorageFormat(object grainState, GrainStateEntity entity)
 
                 if (Log.IsVerbose3) Log.Verbose3("Writing binary data size = {0} for grain id = Partition={1} / Row={2}",
                     data.Length, entity.PartitionKey, entity.RowKey);
-                
-                dataSize = data.Length;
-                entity.Data = data;
+
+                CheckMaxDataSize(data.Length, MAX_DATA_CHUNK_SIZE * MAX_DATA_CHUNKS_COUNT);
+
+                foreach (var partition in SplitBinaryData(data).Zip(GetBinaryDataPropertyNames(), (chunk, name) => new { chunk, name }))
+                {
+                    entity.Properties.Add(partition.name, new EntityProperty(partition.chunk));
+                }
             }
-            if (dataSize > MAX_DATA_SIZE)
+        }
+
+        private void CheckMaxDataSize(int dataSize, int maxDataSize)
+        {
+            if (dataSize > maxDataSize)
             {
-                var msg = string.Format("Data too large to write to Azure table. Size={0} MaxSize={1}", dataSize, MAX_DATA_SIZE);
+                var msg = string.Format("Data too large to write to Azure table. Size={0} MaxSize={1}", dataSize, maxDataSize);
                 Log.Error(0, msg);
                 throw new ArgumentOutOfRangeException("GrainState.Size", msg);
             }
         }
 
+        private static IEnumerable<byte[]> SplitBinaryData(byte[] binaryData)
+        {
+            var chunkCount = (binaryData.Length - 1) / MAX_DATA_CHUNK_SIZE + 1;
+            for (var i = 0; i < chunkCount; i++)
+            {
+                yield return binaryData.Skip(i * MAX_DATA_CHUNK_SIZE).Take(MAX_DATA_CHUNK_SIZE).ToArray();
+            }
+        }
+
+        private static IEnumerable<string> GetBinaryDataPropertyNames()
+        {
+            yield return BINARY_DATA_PROPERTY_NAME;
+            for (var i = 1; i < MAX_DATA_CHUNKS_COUNT; ++i)
+            {
+                yield return BINARY_DATA_PROPERTY_NAME + i;
+            }
+        }
+
+        private IEnumerable<byte[]> ReadBinaryDataChunks(DynamicTableEntity entity)
+        {
+            foreach (var binaryDataPropertyName in GetBinaryDataPropertyNames())
+            {
+                EntityProperty dataProperty;
+                if (entity.Properties.TryGetValue(binaryDataPropertyName, out dataProperty))
+                {
+                    var data = dataProperty.BinaryValue;
+                    if (data != null && data.Length > 0)
+                    {
+                        yield return data;
+                    }
+                }
+            }
+        }
+
+        private byte[] TryReadBinaryData(DynamicTableEntity entity)
+        {
+            return ReadBinaryDataChunks(entity)
+                .Aggregate(Enumerable.Empty<byte>(), (current, binaryDataChunk) => current.Concat(binaryDataChunk))
```

Comment:
I think this Aggregate and concat might be very in-efficient. Anyway, even if it is not, I personalty can't read that. It is like deciphering a puzzle for me. Code needs to be self explanatory. Please rewrite without LINQ, with simple loop. Not cool, but maintainable.